File: 09/intcode-computer.js
const fs = require('fs');

class IntcodeComputer {
    constructor(programCode) {
        this.intcodeSeq = this.loadProgram(programCode)

        this.inputs = [];
        this.outputs = [];

        this.nextJump = 0;
        this.currPos = 0;
        this.relativeBase = 0;
    }

    loadProgram(programCode) {
        // if input is already an array return it, else read the program file
        if (typeof(programCode) == "object") {
            return programCode
        } else {
            return fs.readFileSync(programCode).toString().split(",").map(elem => parseInt(elem));
        }
    }

    extendMemory(memoryAddress) {
        if (memoryAddress >= this.intcodeSeq.length) {
            let memoryEnd = this.intcodeSeq.length

            this.intcodeSeq.length = memoryAddress + 1;
            this.intcodeSeq.fill(0, memoryEnd);
        }
    }

    getParamsValues(params) {
        // Keep only opcode modes
        let modes = Math.trunc(this.intcodeSeq[this.currPos] / 100).toString().split('').reverse();

        let nbParam = params.length

        // add missing modes
        while (nbParam > modes.length) {
            modes.push(0);
        }

        // Get the values
        let values = [];
        for (let i = 0; i < nbParam; i++) {
            let valuePos;

            // position mode (get value at the position of param)
            if (modes[i] == 0) {
                valuePos = this.intcodeSeq[this.currPos + i + 1];
            }
            // immediate mode
            else if (modes[i] == 1) {
                valuePos = this.currPos + i + 1;
            }
            // relative mode
            else if (modes[i] == 2) {
                valuePos = this.relativeBase + this.intcodeSeq[this.currPos + i + 1];
            }

            // Extend memory if needed
            this.extendMemory(valuePos)

            // add parameters to values
            if (params[i] === 'v'){
                values.push(this.intcodeSeq[valuePos]);
            }
            else {
                values.push(valuePos);
            }
        }

        return values;
    }

    run() {
        let values;

        do {
            this.currPos += this.nextJump;

            let opcode = this.intcodeSeq[this.currPos] % 100;

            switch(opcode) {
                case 1:
                    // add: Add param1 and param2, store in param3
                    values = this.getParamsValues(['v', 'v', 'p']);
                    this.intcodeSeq[values[2]] = values[0] + values[1];
                    this.nextJump = 4;
                    break;
                case 2:
                    // multiply: Multiply param1 and param2, store in param3
                    values = this.getParamsValues(['v', 'v', 'p']);
                    this.intcodeSeq[values[2]] = values[0] * values[1];
                    this.nextJump = 4;
                    break;
                // input: Set input value to first param address
                case 3:
                    values = this.getParamsValues(['p']);
                    this.intcodeSeq[values[0]] = this.inputs.shift();
                    this.nextJump = 2;
                    break;
                // output: Output value at first param address
                case 4:
                    values = this.getParamsValues(['v']);
                    console.log(values[0])
                    this.outputs.push(values[0])
                    this.nextJump = 2;
                    break;
                // jump-if-true: jump to second param address if first param != 0
                case 5:
                    values = this.getParamsValues(['v', 'v']);
                    if (values[0]) {
                        this.currPos = values[1];
                        this.nextJump = 0;
                    } else {
                        this.nextJump = 3
                    }
                    break;
                // jump-if-false: jump to second param address if first param === 0
                case 6:
                    values = this.getParamsValues(['v', 'v']);
                    if (!values[0]) {
                        this.currPos = values[1];
                        this.nextJump = 0;
                    } else {
                        this.nextJump = 3;
                    }
                    break;
                // less than: if param1 < param2 store 1 in param3 else store 0 in param3
                case 7:
                    values = this.getParamsValues(['v', 'v', 'p']);
                    if (values[0] < values[1]) {
                        this.intcodeSeq[values[2]] = 1;
                    } else {
                        this.intcodeSeq[values[2]] = 0;
                    }
                    this.nextJump = 4;
                    break;
                // equals: if param1 = param2 store 1 in param3 else store 0 in param3
                case 8:
                    values = this.getParamsValues(['v', 'v', 'p']);
                    if (values[0] === values[1]) {
                        this.intcodeSeq[values[2]] = 1;
                    } else {
                        this.intcodeSeq[values[2]] = 0;
                    }
                    this.nextJump = 4;
                    break;
                case 9:
                    this.relativeBase += this.getParamsValues(['v'])[0];
                    this.nextJump = 2;
                    break;
                case 99:
                    return 'halt';
                default:
                    console.log("Something went wrong");
                    console.log("Debug info:");
                    console.log(`current position: ${this.currPos}`);
                    console.log(`attempted opcode: ${opcode}`);
                    console.log(`previous position: ${this.currPos - this.nextJump}`);
                    console.log(`previous opcode: ${this.intcodeSeq[this.currPos - this.nextJump]}`);
                    process.exit(1);
              }

        } while (this.currPos <= this.intcodeSeq.length)
    }
}

module.exports = IntcodeComputer;

File: 23/intcode_computer.py
#!/usr/bin/env python3

import itertools
import logging
import math
import sys
import time

class IntcodeComputer:
    def __init__(self, program_code, pause_on_output = False, logger = logging):
        # Set logging level
        logger.getLogger().setLevel(logging.INFO)
        self.logger = logger

        self.intcode_seq = self.load_program(program_code)
        self.logger.debug(f'program loaded: {self.intcode_seq}')

        self.pause_on_output = pause_on_output

        self.inputs = []
        self.outputs = []

        self.next_jump = 0
        self.curr_pos = 0
        self.relative_base = 0

        self.state = 'READY'
        self.pause = False

        # Value to use if inputs is empty and opcode 3 is called
        self.default_input = -1

        self.opcode_func = {
            1: self.add,
            2: self.multiply,
            3: self.input_val,
            4: self.output,
            5: self.jump_if_true,
            6: self.jump_if_false,
            7: self.less_than,
            8: self.equals,
            9: self.change_rel_base,
            99: self.halt,
        }


    def load_program(self, program_code):
        self.logger.debug("Loading program:")
        if type(program_code) == str:
            return list(
                map(
                    lambda e: int(e),
                    open(program_code).read().split(',')
                )
            )
        elif type(program_code) == list:
            return program_code
        else:
            print("Invalid program format")
            exit(1)


    def extend_memory(self, memory_address):
        if memory_address >= len(self.intcode_seq):
            self.intcode_seq.extend(
                itertools.repeat(0, memory_address - len(self.intcode_seq) + 1)
            )


    def get_params(self, params):
        # Get modes
        modes = list(
            map(
                lambda e: int(e),
                list(str(math.trunc(self.intcode_seq[self.curr_pos] / 100)))
            )
        )
        modes.reverse()

        nb_param = len(params)

        while nb_param > len(modes):
            modes.append(0)

        values = []
        for i in range(nb_param):
            value_pos = 0
            # Position mode (get value at the position of param)
            if modes[i] == 0:
                value_pos = self.intcode_seq[self.curr_pos + i + 1]
            # Immediate mode
            elif modes[i] == 1:
                value_pos = self.curr_pos + i + 1
            # Relative mode
            elif modes[i] == 2:
                value_pos = self.relative_base + self.intcode_seq[self.curr_pos + i + 1]

            # Extend memory if needed
            self.extend_memory(value_pos)

            # Add value or position
            if params[i] == 'v':
                values.append(self.intcode_seq[value_pos])
            elif params[i] == 'p':
                values.append(value_pos)

        return values


    def add(self):
        values = self.get_params(['v', 'v', 'p'])
        self.logger.debug(f'values: {values}')
        self.logger.debug(f'opcode add: setting position {values[2]} to {values[0]} + {values[1]} ({values[0] + values[1]})')
        self.intcode_seq[values[2]] = values[0] + values[1]
        self.next_jump = 4


    def multiply(self):
        values = self.get_params(['v', 'v', 'p'])
        self.logger.debug(f'values: {values}')
        self.logger.debug(f'opcode multiply: setting position {values[2]} to {values[0]} * {values[1]} ({values[0] * values[1]})')
        self.intcode_seq[values[2]] = values[0] * values[1]
        self.next_jump = 4


    def input_val(self):
        values = self.get_params(['p'])
        input_value = self.inputs.pop(0) if self.inputs else self.default_input

        # Keeps the computer from using too much ressources while idle
        if input_value == -1:
            time.sleep(1)

        self.logger.debug(f'values: {values}')
        self.logger.debug(f'opcode input: putting {input_value} in position {values[0]}')
        self.intcode_seq[values[0]] = input_value
        self.next_jump = 2


    def output(self):
        value = self.get_params(['v'])[0]
        self.logger.debug(f'opcode output: outputting {value}')
        self.outputs.append(value)
        self.next_jump = 2

        if self.pause_on_output:
            self.logger.debug('=== pause on output is on, returning')
            # Will pause after this opcode
            self.pause = True


    def jump_if_true(self):
        values = self.get_params(['v', 'v'])

        if values[0] != 0:
            self.logger.debug(f'values: {values}')
            self.logger.debug(f'opcode jump_if_true: {values[0]} is not 0, jumping to {values[1]}')
            self.curr_pos = values[1]
            self.next_jump = 0
        else:
            self.logger.debug(f'values: {values}')
            self.logger.debug(f'opcode jump_if_true: {values[0]} is 0, NOT jumping to {values[1]}')
            self.next_jump = 3


    def jump_if_false(self):
        values = self.get_params(['v', 'v'])

        if values[0] == 0:
            self.logger.debug(f'values: {values}')
            self.logger.debug(f'opcode jump_if_false: {values[0]} is 0, jumping to {values[1]}')
            self.curr_pos = values[1]
            self.next_jump = 0
        else:
            self.logger.debug(f'values: {values}')
            self.logger.debug(f'opcode jump_if_false: {values[0]} is not 0, NOT jumping to {values[1]}')
            self.next_jump = 3


    def less_than(self):
        values = self.get_params(['v', 'v', 'p'])

        if values[0] < values[1]:
            self.logger.debug(f'values: {values}')
            self.logger.debug(f'opcode less_than: {values[0]} is less than {values[1]} setting position {values[2]} to 1')
            self.intcode_seq[values[2]] = 1
        else:
            self.logger.debug(f'values: {values}')
            self.logger.debug(f'opcode less_than: {values[0]} is NOT less than {values[1]} setting position {values[2]} to 0')
            self.intcode_seq[values[2]] = 0

        self.next_jump = 4


    def equals(self):
        values = self.get_params(['v', 'v', 'p'])

        if values[0] == values[1]:
            self.logger.debug(f'values: {values}')
            self.logger.debug(f'opcode equals: {values[0]} is equal to {values[1]} setting position {values[2]} to 1')
            self.intcode_seq[values[2]] = 1
        else:
            self.logger.debug(f'values: {values}')
            self.logger.debug(f'opcode equals: {values[0]} is NOT equal to {values[1]} setting position {values[2]} to 0')
            self.intcode_seq[values[2]] = 0

        self.next_jump = 4


    def change_rel_base(self):
        new_rel_base = self.relative_base + self.get_params(['v'])[0]
        self.logger.debug(f'opcode change_rel_base: changing relative base from {self.relative_base} to {new_rel_base}')
        self.relative_base = new_rel_base
        self.next_jump = 2


    def halt(self):
        self.logger.debug(f'opcode halt: halting program')
        self.state = 'HALTED'
        self.pause = True


    def run(self):
        self.pause = False

        if self.state == 'READY' or 'PAUSED':
            self.state = 'RUNNING'
        else:
            print(f'cannot run in state {self.state}')

        while (self.curr_pos != len(self.intcode_seq)):
            self.curr_pos = self.curr_pos + self.next_jump


            opcode = self.intcode_seq[self.curr_pos] % 100

            self.logger.debug(f'current position: {self.curr_pos}')
            self.logger.debug(f'opcode: {opcode}')
            self.logger.debug(f'full opcode: {self.intcode_seq[self.curr_pos]}')

            # Execute opcode
            self.opcode_func[opcode]()

            if self.pause:
                self.logger.debug('pause was triggered, pausing')
                return
